### DISCUSSION TITLE
Ethereum - User operations EIP-4337. All fields MUST be set as hex values.

### DISCUSSION THREAD

**Discussion by DenisDemyanko** [trustwallet/wallet-core#3366] [Q&A]
Hello, please help me. I want to receive JSON from user operations EIP-4337.

I received JSON, but JSON does not contain hex values (gas values & nonce)

When I try sent this JSON, I received the error - "Invalid hex value for property nonce:0 in UserOp" or  "Invalid hex value for property maxFeePerGas:1507969922 in UserOp".

Please help me, how I can get JSON with hex values. Or what do I need to receive valid EthereumSigningOutput.  
And I think the signature is invalid because we do sign not hex values? Or maybe you can help me, with how to receive a valid signature, if I will change all values to hex in JSON. 

[UserOperation a full user-operation struct. All fields MUST be set as hex values. empty bytes block (e.g. empty initCode) MUST be set to "0x"](https://eips.ethereum.org/EIPS/eip-4337#parameters)

My code:
`let input = EthereumSigningInput.with {
            $0.chainID = Data(hexString: "0x000539")!
            $0.nonce = Data(hexString: "00")!
            $0.maxFeePerGas = Data(hexString: "0x59e1cb82")!
            $0.maxInclusionFeePerGas = Data(hexString: "0x59e1cb82")! //maxPriorityFeePerGas
            $0.txMode = .userOp
            $0.gasLimit = Data(hexString: "0x5454")!
            $0.privateKey = wallet.getKeyForCoin(coin: .ethereum).data
            $0.userOperation = EthereumUserOperation.with {
                $0.verificationGasLimit = Data(hexString: "0x05f997")!
                $0.preVerificationGas = Data(hexString: "0x015bd8")!
                $0.entryPoint = entryPointAddress
                $0.sender = myWalletTokenAddress
                $0.initCode = Data(hexString: initCode)!
                $0.paymasterAndData = Data(hexString: defaultPaymaster)!
            }
        }`

`let output: EthereumSigningOutput = AnySigner.sign(input: input, coin: .ethereum)
`

`let jsonString = String(data: output.encoded, encoding: .utf8)`

Example of JSON:

`{"callData":"0xb61d27f60000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000000","callGasLimit":"21588","initCode":"0x40154c8909789bef6e8ed93befd234334c8d11155fbfb9cf000000000000000000000000b40284e302dc2d6e9e7be7784ba0ebe4aef7e5b60000000000000000000000000000000000000000000000000000000000000000","maxFeePerGas":"1507969922","maxPriorityFeePerGas":"1507969922","nonce":"0","paymasterAndData":"0xd0961c1ce3c333040ba3630fc64bf3ea84fef1bf","preVerificationGas":"89048","sender":"0x6d7185f1a41179fc8788a8aed97218d2403375bf","signature":"0x2276b6ce2fbe379d9761177174cdda9e9cfa001d23dd9c31d89c579a9d99e2450135c4f481aa9c44bbe1c52b57461b70ff7ec657931c363131c5d6ce8b086df51b","verificationGasLimit":"391575"}`

**Comment by satoshiotomakan** [comment]
Hi @DenisDemyanko, you can try to use [java.math.BigInteger](https://www.geeksforgeeks.org/biginteger-class-in-java/) to parse the decimal string and convert to a hex string:
```kt
BigInteger maxFeePerGas = new BigInteger(receivedJson.maxFeePerGas);
let input = EthereumSigningInput.with {
    $0.maxFeePerGas = Data(hexString: maxFeePerGas.toString(16))!
    // ...
};
```

**Comment by DenisDemyanko** [comment]
@satoshiotomakan thank you.
Could you please help me, what am I doing wrong? I am getting the wrong signature.
I want to call **_eth_sendUserOperation_** method, but when I try to send method, I received error: 
```
 "error": {
        "message": "Invalid UserOp signature or paymaster signature",
        "code": -32507
    }
```

The pre-hash of the transaction is valid, because if I manually change the signature in the received JSON to valid signature, the result of the API function will be a hash that matches **`output.preHash.hexString`**

```
{
    "jsonrpc": "2.0",
    "id": 44,
    "result": "0x8321b2d6301b0f49945eabbfd38f7dcaa53ac846933468a165f0695500f3f756"
}
```

**Here is my code to generate output**

```
`let factoryAddress = "0x40154C8909789beF6e8ED93bEFD234334C8d1115"
let defaultPaymaster = "0xd0961c1Ce3c333040BA3630FC64Bf3ea84FeF1bf"
let deafultToken = "0x46aBf0F64103c8B08d6D8b4663261D83595fBE97"
let entryPointAddress = "0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789"
        
let mySmartWalletAddress = "0x32480A1e937b7749a755793eA0F3D9607052a91A"
let myPublicWalletAddress = "0x199b0A98F87622bDB4B8Eaeb05171c86652dCa10"
        
let function = EthereumAbiFunction(name: "createAccount")
function.addParamAddress(val: Data (hexString: myPublicWalletAddress)!, isOutput: false)
function.addParamUInt256(val: Data(hexString: "00")!, isOutput: false)
        
let createAccountFuncEncoded = EthereumAbi.encode(fn: function)
let initCode = factoryAddress + createAccountFuncEncoded.hexString
        
let maxFeePerGas = "0x59818264"
let maxPriorityFeePerGas = "0x59682f00"
let gasLimit = "0x5454"
let verificationGasLimit = "0x05f997"
let preVerificationGas = "0x015bf0"
let nonce = "0x00"
let chainID = "0x0539"
let privateKey = "0x220b646d64f847d5c807fc37db2511b140406a8c1d4a3580b8b29ba980063dab"

let input = EthereumSigningInput.with {
            $0.chainID = Data(hexString: chainID)!
            $0.nonce = Data(hexString: nonce)!
            $0.maxFeePerGas = Data(hexString: maxFeePerGas)!
            $0.maxInclusionFeePerGas = Data(hexString: maxPriorityFeePerGas)!
            $0.txMode = .userOp
            $0.gasLimit = Data(hexString: gasLimit)!
            $0.privateKey = Data(hexString: privateKey)!
            $0.userOperation = EthereumUserOperation.with {
                $0.verificationGasLimit = Data(hexString: verificationGasLimit)!
                $0.preVerificationGas = Data(hexString: preVerificationGas)!
                $0.entryPoint = entryPointAddress
                $0.sender = mySmartWalletAddress
                $0.initCode = Data(hexString: initCode)!
                $0.paymasterAndData = Data(hexString: defaultPaymaster)!
            }
        }`
```

```
let output: EthereumSigningOutput = AnySigner.sign(input: input, coin: .ethereum) 
let jsonString = String(data: output.encoded, encoding: .utf8) ?? ""
print(jsonString)
```

**JSON result:**

`{"callData":"0xb61d27f60000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000000","callGasLimit":"21588","initCode":"0x40154c8909789bef6e8ed93befd234334c8d11155fbfb9cf000000000000000000000000199b0a98f87622bdb4b8eaeb05171c86652dca100000000000000000000000000000000000000000000000000000000000000000","maxFeePerGas":"1501659748","maxPriorityFeePerGas":"1500000000","nonce":"0","paymasterAndData":"0xd0961c1ce3c333040ba3630fc64bf3ea84fef1bf","preVerificationGas":"89072","sender":"0x32480a1e937b7749a755793ea0f3d9607052a91a","signature":"0x81710570a173bbd96413e4558e5fed7adede220abd223dd87b665d3097b04f9b528fb32d1ee822db337117df5e58659f78ab2dff285750c15d0a29c5e0a678d71c","verificationGasLimit":"391575"}`

**API Call**

```
{"method":"eth_sendUserOperation","params":[
{
  "callData" : "0xb61d27f60000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000000",
  "paymasterAndData" : "0xd0961c1ce3c333040ba3630fc64bf3ea84fef1bf",
  "preVerificationGas" : "0x015bf0",
  "maxPriorityFeePerGas" : "0x59682f00",
  "sender" : "0x32480a1e937b7749a755793ea0f3d9607052a91a",
  "signature" : "0x81710570a173bbd96413e4558e5fed7adede220abd223dd87b665d3097b04f9b528fb32d1ee822db337117df5e58659f78ab2dff285750c15d0a29c5e0a678d71c",
  "verificationGasLimit" : "0x05f997",
  "maxFeePerGas" : "0x59818264",
  "callGasLimit" : "0x5454",
  "initCode" : "0x40154c8909789bef6e8ed93befd234334c8d11155fbfb9cf000000000000000000000000199b0a98f87622bdb4b8eaeb05171c86652dca100000000000000000000000000000000000000000000000000000000000000000",
  "nonce" : "0x00"
},"0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789"],"id":44,"jsonrpc":"2.0"}
```


**The expected valid signature should be** 
**0x87239c9d96d1199f1b4bb1bb7804cd5c16ca2ba875c5767f802b938f209aa3ce289c08f44a9d0c09aa49aeee81ce1435a3153295163aca4902de6a9b557678a61c**

I know the valid signature because I have a browser extension, that uses the @account-abstraction/contracts package.
This browser extension makes calls and I can debug and compile logs. All data is the same, except for the signature.

Maybe I missed some parameters when generation input? But pre-hash is the same and valid.

Thank you.

**Comment by DenisDemyanko** [comment]
@satoshiotomakan

Hello. Also, I try node lib. Same result.
I got a valid preHash but an invalid signature.
If I sign the hash via Ethereum lib I get a valid signature.

Signature from core-wallet - **0x6591fdd130a3dc87f3b09cd0fe55edae90ccce1e306dce51c648f1742026fdb52f3c6287c0912385852c524bfedff6d53fe76a2b7e8cb672e03de4f66b6237061b**

Signature from ETH - **0xbc13fcd6f072acd8399356018999e730f0c108ac590f4f76900ec7637417d663138f422f75cd3a277a766d9482b18e6752343bcf8785356c7d1e27b5f1f82db01c**

Prehash - **0x239e8ae401f75a59a655e782745966a9b8d559f57e02a581cf09516f019a4eac**

```
import { TW } from "@trustwallet/wallet-core"
import { initWasm } from "@trustwallet/wallet-core"
import { ethers } from "ethers";

let func = async () => {
    const { AnySigner, CoinType, HexCoding } = await initWasm();

    const input = TW.Ethereum.Proto.SigningInput.create({
        txMode: TW.Ethereum.Proto.TransactionMode.UserOp,
        chainId: Buffer.from("0539", "hex"),
        nonce: Buffer.from("00", "hex"),
        maxFeePerGas: Buffer.from("596ffcba"  , "hex"),
        maxInclusionFeePerGas: Buffer.from("59682f00"  , "hex"),
        gasLimit: Buffer.from("5454", "hex"),
        userOperation: TW.Ethereum.Proto.UserOperation.create({
            entryPoint: "0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789",
            initCode: Buffer.from("40154c8909789bef6e8ed93befd234334c8d11155fbfb9cf00000000000000000000000094d773ecc236473bb2cccc3d7b4f1731e1f305980000000000000000000000000000000000000000000000000000000000000000", "hex"),
            sender: "0x20c008674D5e1AFde073A2A51A4A960107e3c9D7",
            preVerificationGas: Buffer.from("015bf0", "hex"),
            verificationGasLimit: Buffer.from("05f997", "hex"),
            paymasterAndData: Buffer.from("d0961c1Ce3c333040BA3630FC64Bf3ea84FeF1bf", "hex")
        }),
        privateKey: HexCoding.decode(
          "0xb9db2151034c6b6604d6aaf86ef0852e5c662a456b2365aab389714d03e9168e"
        ),
      });

      input.txMode = TW.Ethereum.Proto.TransactionMode.UserOp
    
      const encoded = TW.Ethereum.Proto.SigningInput.encode(input).finish();
      const outputData = AnySigner.sign(encoded, CoinType.ethereum);
      const output = TW.Ethereum.Proto.SigningOutput.decode(outputData);
      console.log(String.fromCharCode(...output.encoded))
      console.log(HexCoding.encode(output.preHash))

      let wallet = new ethers.Wallet("0xb9db2151034c6b6604d6aaf86ef0852e5c662a456b2365aab389714d03e9168e")

      let res = await wallet.signMessage(output.preHash)
      console.log(res)

} 

func()

```

**Reply by satoshiotomakan** [reply]
As I can see, you provide different private keys:
```
let wallet = new ethers.Wallet("0xb9db2151034c6b6604d6aaf86ef0852e5c662a456b2365aab389714d03e9168e")
```
```
let privateKey = "0x220b646d64f847d5c807fc37db2511b140406a8c1d4a3580b8b29ba980063dab"
```

Cat it be the reason?

**Reply by DenisDemyanko** [reply]
@satoshiotomakan  No, because the code from Swift and TypeScript uses different private keys. It's okay.
The main problem or question for example (TypeScript code above this post) is that signatures generated from TW and ETH are different. The pre-hash is valid. If I sign the pre-hash with an ETH wallet, I get a valid signature that works.
The signature created via TW not working and I get an error - _"message": "Invalid UserOp signature or paymaster signature"_
```
let wallet = new ethers.Wallet("0xb9db2151034c6b6604d6aaf86ef0852e5c662a456b2365aab389714d03e9168e")
let res = await wallet.signMessage(output.preHash)
```

I provided TypeScript code, so you can run in the terminal and see signatures that return from TW and ETH.
[TW_TEST.zip](https://github.com/trustwallet/wallet-core/files/12343906/TW_TEST.zip)

```
- yarn
- yarn run start
```

Result:

```
`OUTPUT TW:
{"callData":"0xb61d27f60000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000000","callGasLimit":"21588","initCode":"0x40154c8909789bef6e8ed93befd234334c8d11155fbfb9cf00000000000000000000000094d773ecc236473bb2cccc3d7b4f1731e1f305980000000000000000000000000000000000000000000000000000000000000000","maxFeePerGas":"1500511418","maxPriorityFeePerGas":"1500000000","nonce":"0","paymasterAndData":"0xd0961c1ce3c333040ba3630fc64bf3ea84fef1bf","preVerificationGas":"89072","sender":"0x20c008674d5e1afde073a2a51a4a960107e3c9d7","signature":"0x6591fdd130a3dc87f3b09cd0fe55edae90ccce1e306dce51c648f1742026fdb52f3c6287c0912385852c524bfedff6d53fe76a2b7e8cb672e03de4f66b6237061b","verificationGasLimit":"391575"}

PRE-HASH TW:0x239e8ae401f75a59a655e782745966a9b8d559f57e02a581cf09516f019a4eac

SIGANTURE FROM ETH: 0xbc13fcd6f072acd8399356018999e730f0c108ac590f4f76900ec7637417d663138f422f75cd3a277a766d9482b18e6752343bcf8785356c7d1e27b5f1f82db01c`
```
